File: projects/Dotproduct.java
import java.util.Locale;
import java.util.Random;

public final class Dotproduct {
    private static final int[] SIZES = new int[] {10, 100, 1_000, 10_000, 100_000, 1_000_000};

    private static volatile long checksumInt = 0;
    private static volatile double checksumFp = 0.0;

    private static int repsForN(int n) {
        // Mirror the C-side policy: target ~20M multiply-adds.
        long targetOps = 20_000_000L;
        long reps = targetOps / Math.max(1, n);
        if (reps < 3) reps = 3;
        if (reps > 2_000_000L) reps = 2_000_000L;
        return (int) reps;
    }

    private static void printHeader() {
        System.out.println("lang,type,N,reps,total_ns,avg_ns,checksum");
    }

    private static void fill(byte[] a, byte[] b, int n, long seed) {
        Random r = new Random(seed);
        for (int i = 0; i < n; i++) {
            // Small range [-15, 15].
            a[i] = (byte) (r.nextInt(31) - 15);
            b[i] = (byte) (r.nextInt(31) - 15);
        }
    }

    private static void fill(short[] a, short[] b, int n, long seed) {
        Random r = new Random(seed);
        for (int i = 0; i < n; i++) {
            a[i] = (short) (r.nextInt(2001) - 1000);
            b[i] = (short) (r.nextInt(2001) - 1000);
        }
    }

    private static void fill(int[] a, int[] b, int n, long seed) {
        Random r = new Random(seed);
        for (int i = 0; i < n; i++) {
            a[i] = r.nextInt(2_000_001) - 1_000_000;
            b[i] = r.nextInt(2_000_001) - 1_000_000;
        }
    }

    private static void fill(float[] a, float[] b, int n, long seed) {
        Random r = new Random(seed);
        for (int i = 0; i < n; i++) {
            // Roughly [-1, 1).
            a[i] = (float) (r.nextDouble() * 2.0 - 1.0);
            b[i] = (float) (r.nextDouble() * 2.0 - 1.0);
        }
    }

    private static void fill(double[] a, double[] b, int n, long seed) {
        Random r = new Random(seed);
        for (int i = 0; i < n; i++) {
            a[i] = r.nextDouble() * 2.0 - 1.0;
            b[i] = r.nextDouble() * 2.0 - 1.0;
        }
    }

    private static long dotOnce(byte[] a, byte[] b, int n) {
        long acc = 0;
        for (int i = 0; i < n; i++) acc += (long) a[i] * (long) b[i];
        return acc;
    }

    private static long dotOnce(short[] a, short[] b, int n) {
        long acc = 0;
        for (int i = 0; i < n; i++) acc += (long) a[i] * (long) b[i];
        return acc;
    }

    private static long dotOnce(int[] a, int[] b, int n) {
        long acc = 0;
        for (int i = 0; i < n; i++) acc += (long) a[i] * (long) b[i];
        return acc;
    }

    private static double dotOnce(float[] a, float[] b, int n) {
        double acc = 0.0;
        for (int i = 0; i < n; i++) acc += (double) a[i] * (double) b[i];
        return acc;
    }

    private static double dotOnce(double[] a, double[] b, int n) {
        double acc = 0.0;
        for (int i = 0; i < n; i++) acc += a[i] * b[i];
        return acc;
    }

    private static void warmupByte() {
        int n = 100_000;
        byte[] a = new byte[n];
        byte[] b = new byte[n];
        fill(a, b, n, 0xC0FFEE + n);
        for (int i = 0; i < 10; i++) checksumInt += dotOnce(a, b, n);
    }

    private static void warmupShort() {
        int n = 100_000;
        short[] a = new short[n];
        short[] b = new short[n];
        fill(a, b, n, 0xBADC0DE + n);
        for (int i = 0; i < 10; i++) checksumInt += dotOnce(a, b, n);
    }

    private static void warmupInt() {
        int n = 100_000;
        int[] a = new int[n];
        int[] b = new int[n];
        fill(a, b, n, 0x12345678L + n);
        for (int i = 0; i < 10; i++) checksumInt += dotOnce(a, b, n);
    }

    private static void warmupFloat() {
        int n = 100_000;
        float[] a = new float[n];
        float[] b = new float[n];
        fill(a, b, n, 0x31415926L + n);
        for (int i = 0; i < 10; i++) checksumFp += dotOnce(a, b, n);
    }

    private static void warmupDouble() {
        int n = 100_000;
        double[] a = new double[n];
        double[] b = new double[n];
        fill(a, b, n, 0x27182818L + n);
        for (int i = 0; i < 10; i++) checksumFp += dotOnce(a, b, n);
    }

    private static void benchByte(int n) {
        byte[] a = new byte[n];
        byte[] b = new byte[n];
        fill(a, b, n, 0xC0FFEE + n);

        int reps = repsForN(n);
        long sum = 0;
        long t0 = System.nanoTime();
        for (int r = 0; r < reps; r++) sum += dotOnce(a, b, n);
        long t1 = System.nanoTime();

        checksumInt += sum;
        long total = t1 - t0;
        double avg = (double) total / (double) reps;
        System.out.printf(Locale.ROOT, "Java,signed_char,%d,%d,%d,%.2f,%d%n",
                n, reps, total, avg, checksumInt);
    }

    private static void benchShort(int n) {
        short[] a = new short[n];
        short[] b = new short[n];
        fill(a, b, n, 0xBADC0DE + n);

        int reps = repsForN(n);
        long sum = 0;
        long t0 = System.nanoTime();
        for (int r = 0; r < reps; r++) sum += dotOnce(a, b, n);
        long t1 = System.nanoTime();

        checksumInt += sum;
        long total = t1 - t0;
        double avg = (double) total / (double) reps;
        System.out.printf(Locale.ROOT, "Java,short,%d,%d,%d,%.2f,%d%n",
                n, reps, total, avg, checksumInt);
    }

    private static void benchInt(int n) {
        int[] a = new int[n];
        int[] b = new int[n];
        fill(a, b, n, 0x12345678L + n);

        int reps = repsForN(n);
        long sum = 0;
        long t0 = System.nanoTime();
        for (int r = 0; r < reps; r++) sum += dotOnce(a, b, n);
        long t1 = System.nanoTime();

        checksumInt += sum;
        long total = t1 - t0;
        double avg = (double) total / (double) reps;
        System.out.printf(Locale.ROOT, "Java,int,%d,%d,%d,%.2f,%d%n",
                n, reps, total, avg, checksumInt);
    }

    private static void benchFloat(int n) {
        float[] a = new float[n];
        float[] b = new float[n];
        fill(a, b, n, 0x31415926L + n);

        int reps = repsForN(n);
        double sum = 0.0;
        long t0 = System.nanoTime();
        for (int r = 0; r < reps; r++) sum += dotOnce(a, b, n);
        long t1 = System.nanoTime();

        checksumFp += sum;
        long total = t1 - t0;
        double avg = (double) total / (double) reps;
        System.out.printf(Locale.ROOT, "Java,float,%d,%d,%d,%.2f,%.6f%n",
                n, reps, total, avg, checksumFp);
    }

    private static void benchDouble(int n) {
        double[] a = new double[n];
        double[] b = new double[n];
        fill(a, b, n, 0x27182818L + n);

        int reps = repsForN(n);
        double sum = 0.0;
        long t0 = System.nanoTime();
        for (int r = 0; r < reps; r++) sum += dotOnce(a, b, n);
        long t1 = System.nanoTime();

        checksumFp += sum;
        long total = t1 - t0;
        double avg = (double) total / (double) reps;
        System.out.printf(Locale.ROOT, "Java,double,%d,%d,%d,%.2f,%.6f%n",
                n, reps, total, avg, checksumFp);
    }

    public static void main(String[] args) {
        printHeader();

        // Warm-up to encourage JIT compilation before real measurements.
        warmupByte();
        warmupShort();
        warmupInt();
        warmupFloat();
        warmupDouble();

        for (int n : SIZES) benchByte(n);
        for (int n : SIZES) benchShort(n);
        for (int n : SIZES) benchInt(n);
        for (int n : SIZES) benchFloat(n);
        for (int n : SIZES) benchDouble(n);

        System.err.printf(Locale.ROOT, "checksum_int=%d checksum_fp=%.6f%n", checksumInt, checksumFp);
    }
}

File: projects/dotproduct.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <inttypes.h>
#include <time.h>

// Fixed benchmark sizes (per assignment plan).
static const size_t k_sizes[] = {10, 100, 1000, 10000, 100000, 1000000};
static const size_t k_num_sizes = sizeof(k_sizes) / sizeof(k_sizes[0]);

static volatile double g_checksum_fp = 0.0;
static volatile long long g_checksum_int = 0;

static inline uint64_t xorshift64(uint64_t *state) {
  uint64_t x = *state;
  x ^= x << 13;
  x ^= x >> 7;
  x ^= x << 17;
  *state = x;
  return x;
}

static inline uint64_t now_ns(void) {
  struct timespec ts;
  // CLOCK_MONOTONIC is a steady clock suitable for benchmarking.
  if (clock_gettime(CLOCK_MONOTONIC, &ts) != 0) {
    return 0;
  }
  return (uint64_t)ts.tv_sec * 1000000000ull + (uint64_t)ts.tv_nsec;
}

static inline int reps_for_n(size_t n) {
  // Keep runtimes reasonable and still smooth out noise.
  // For very small n we need many reps; for large n, few reps.
  const size_t target_ops = 20000000ull;  // ~20M multiply-adds.
  size_t reps = target_ops / (n ? n : 1);
  if (reps < 3) reps = 3;
  if (reps > 2000000ull) reps = 2000000ull;
  return (int)reps;
}

static void fill_schar(signed char *a, signed char *b, size_t n, uint64_t seed) {
  uint64_t st = seed;
  for (size_t i = 0; i < n; i++) {
    // Small range to reduce overflow in accumulator while still varying.
    a[i] = (signed char)((int)(xorshift64(&st) % 31) - 15);
    b[i] = (signed char)((int)(xorshift64(&st) % 31) - 15);
  }
}

static void fill_short(short *a, short *b, size_t n, uint64_t seed) {
  uint64_t st = seed;
  for (size_t i = 0; i < n; i++) {
    a[i] = (short)((int)(xorshift64(&st) % 2001) - 1000);
    b[i] = (short)((int)(xorshift64(&st) % 2001) - 1000);
  }
}

static void fill_int(int *a, int *b, size_t n, uint64_t seed) {
  uint64_t st = seed;
  for (size_t i = 0; i < n; i++) {
    a[i] = (int)((int64_t)(xorshift64(&st) % 2000001ull) - 1000000);
    b[i] = (int)((int64_t)(xorshift64(&st) % 2000001ull) - 1000000);
  }
}

static void fill_float(float *a, float *b, size_t n, uint64_t seed) {
  uint64_t st = seed;
  for (size_t i = 0; i < n; i++) {
    // Map to roughly [-1, 1].
    uint64_t ra = xorshift64(&st);
    uint64_t rb = xorshift64(&st);
    a[i] = (float)(((double)(ra & 0xffffu) / 32768.0) - 1.0);
    b[i] = (float)(((double)(rb & 0xffffu) / 32768.0) - 1.0);
  }
}

static void fill_double(double *a, double *b, size_t n, uint64_t seed) {
  uint64_t st = seed;
  for (size_t i = 0; i < n; i++) {
    uint64_t ra = xorshift64(&st);
    uint64_t rb = xorshift64(&st);
    a[i] = ((double)(ra & 0xffffu) / 32768.0) - 1.0;
    b[i] = ((double)(rb & 0xffffu) / 32768.0) - 1.0;
  }
}

static inline long long dot_schar_once(const signed char *a, const signed char *b, size_t n) {
  long long acc = 0;
  for (size_t i = 0; i < n; i++) acc += (long long)a[i] * (long long)b[i];
  return acc;
}

static inline long long dot_short_once(const short *a, const short *b, size_t n) {
  long long acc = 0;
  for (size_t i = 0; i < n; i++) acc += (long long)a[i] * (long long)b[i];
  return acc;
}

static inline long long dot_int_once(const int *a, const int *b, size_t n) {
  long long acc = 0;
  for (size_t i = 0; i < n; i++) acc += (long long)a[i] * (long long)b[i];
  return acc;
}

static inline double dot_float_once(const float *a, const float *b, size_t n) {
  double acc = 0.0;
  for (size_t i = 0; i < n; i++) acc += (double)a[i] * (double)b[i];
  return acc;
}

static inline double dot_double_once(const double *a, const double *b, size_t n) {
  double acc = 0.0;
  for (size_t i = 0; i < n; i++) acc += a[i] * b[i];
  return acc;
}

static void print_csv_header(void) {
  printf("lang,type,N,reps,total_ns,avg_ns,checksum\n");
}

static void bench_schar(size_t n) {
  signed char *a = (signed char *)malloc(n * sizeof(*a));
  signed char *b = (signed char *)malloc(n * sizeof(*b));
  if (!a || !b) {
    fprintf(stderr, "alloc failed for schar N=%zu\n", n);
    exit(1);
  }
  fill_schar(a, b, n, 0xC0FFEEu + (uint64_t)n);

  const int reps = reps_for_n(n);
  long long sum = 0;
  const uint64_t t0 = now_ns();
  for (int r = 0; r < reps; r++) sum += dot_schar_once(a, b, n);
  const uint64_t t1 = now_ns();

  g_checksum_int += sum;
  const uint64_t total = t1 - t0;
  const double avg = (double)total / (double)reps;
  printf("C,signed_char,%zu,%d,%" PRIu64 ",%.2f,%lld\n", n, reps, total, avg, (long long)g_checksum_int);

  free(a);
  free(b);
}

static void bench_short(size_t n) {
  short *a = (short *)malloc(n * sizeof(*a));
  short *b = (short *)malloc(n * sizeof(*b));
  if (!a || !b) {
    fprintf(stderr, "alloc failed for short N=%zu\n", n);
    exit(1);
  }
  fill_short(a, b, n, 0xBADC0DEu + (uint64_t)n);

  const int reps = reps_for_n(n);
  long long sum = 0;
  const uint64_t t0 = now_ns();
  for (int r = 0; r < reps; r++) sum += dot_short_once(a, b, n);
  const uint64_t t1 = now_ns();

  g_checksum_int += sum;
  const uint64_t total = t1 - t0;
  const double avg = (double)total / (double)reps;
  printf("C,short,%zu,%d,%" PRIu64 ",%.2f,%lld\n", n, reps, total, avg, (long long)g_checksum_int);

  free(a);
  free(b);
}

static void bench_int(size_t n) {
  int *a = (int *)malloc(n * sizeof(*a));
  int *b = (int *)malloc(n * sizeof(*b));
  if (!a || !b) {
    fprintf(stderr, "alloc failed for int N=%zu\n", n);
    exit(1);
  }
  fill_int(a, b, n, 0x12345678u + (uint64_t)n);

  const int reps = reps_for_n(n);
  long long sum = 0;
  const uint64_t t0 = now_ns();
  for (int r = 0; r < reps; r++) sum += dot_int_once(a, b, n);
  const uint64_t t1 = now_ns();

  g_checksum_int += sum;
  const uint64_t total = t1 - t0;
  const double avg = (double)total / (double)reps;
  printf("C,int,%zu,%d,%" PRIu64 ",%.2f,%lld\n", n, reps, total, avg, (long long)g_checksum_int);

  free(a);
  free(b);
}

static void bench_float(size_t n) {
  float *a = (float *)malloc(n * sizeof(*a));
  float *b = (float *)malloc(n * sizeof(*b));
  if (!a || !b) {
    fprintf(stderr, "alloc failed for float N=%zu\n", n);
    exit(1);
  }
  fill_float(a, b, n, 0x31415926u + (uint64_t)n);

  const int reps = reps_for_n(n);
  double sum = 0.0;
  const uint64_t t0 = now_ns();
  for (int r = 0; r < reps; r++) sum += dot_float_once(a, b, n);
  const uint64_t t1 = now_ns();

  g_checksum_fp += sum;
  const uint64_t total = t1 - t0;
  const double avg = (double)total / (double)reps;
  printf("C,float,%zu,%d,%" PRIu64 ",%.2f,%.6f\n", n, reps, total, avg, (double)g_checksum_fp);

  free(a);
  free(b);
}

static void bench_double(size_t n) {
  double *a = (double *)malloc(n * sizeof(*a));
  double *b = (double *)malloc(n * sizeof(*b));
  if (!a || !b) {
    fprintf(stderr, "alloc failed for double N=%zu\n", n);
    exit(1);
  }
  fill_double(a, b, n, 0x27182818u + (uint64_t)n);

  const int reps = reps_for_n(n);
  double sum = 0.0;
  const uint64_t t0 = now_ns();
  for (int r = 0; r < reps; r++) sum += dot_double_once(a, b, n);
  const uint64_t t1 = now_ns();

  g_checksum_fp += sum;
  const uint64_t total = t1 - t0;
  const double avg = (double)total / (double)reps;
  printf("C,double,%zu,%d,%" PRIu64 ",%.2f,%.6f\n", n, reps, total, avg, (double)g_checksum_fp);

  free(a);
  free(b);
}

int main(void) {
  print_csv_header();
  for (size_t i = 0; i < k_num_sizes; i++) bench_schar(k_sizes[i]);
  for (size_t i = 0; i < k_num_sizes; i++) bench_short(k_sizes[i]);
  for (size_t i = 0; i < k_num_sizes; i++) bench_int(k_sizes[i]);
  for (size_t i = 0; i < k_num_sizes; i++) bench_float(k_sizes[i]);
  for (size_t i = 0; i < k_num_sizes; i++) bench_double(k_sizes[i]);

  // Ensure side-effects are visible.
  fprintf(stderr, "checksum_int=%lld checksum_fp=%.6f\n", (long long)g_checksum_int, (double)g_checksum_fp);
  return 0;
}
